### ISSUE TITLE
Fossilization and conditioning on survival to present

### ISSUE THREAD

**Issue by wsdewitt** [matsengrp/gcdyn#68]
----

**tl;dr:** _When our likelihood has fossilization, we can't simply discard trees that don't have present-day survivors._

----

At the last meeting, we (@thanasibakis @matsen @vnminin @yss1) spent some time thinking about how the likelihood conditions on survival/sampling, and we weren't 100% sure we understood if we were handling that correctly given our simulation/sampling procedure.

Today @yss1 asked if the way we simulate non-surviving subtrees is consistent with the fact that we discard trees that don't survive to the sampling time. His question led me to think of one concrete way things may be getting messed up:
1. We simulate trees, and don't prune the extinct subtrees or upsampled survivors. We interpret this as giving us trees with a fossilization probability $\sigma=1$ and sampling probability $\rho=1$.
2. We discard trees that don't have survivors at the sampling time, e.g. by sampling trees using [this function](https://github.com/matsengrp/gcdyn/blob/09a1a82601d193db2510797323f6eec295b95a2a/gcdyn/utils.py#L98).
3. We compute likelihoods assuming $\sigma=\rho=1$. [This line](https://github.com/matsengrp/gcdyn/blob/09a1a82601d193db2510797323f6eec295b95a2a/notebooks/inference/birth_death/mcmc/geweke_test.py#L99) suggests that's maybe happening in the Geweke tests.

If this is what we are doing, I think it could be causing problems, as I detail below.

## Probabilistic model assuming observable failures

Recall that our MTBD likelihood is for a partially-observed process, where we only sample the tree subtended by samples and fossils. If a tree has no samples or fossils, the most we can know is that we tried and we got an empty tree. Let $t_\dagger$ denote the empty tree event. In probability space terms, $t_\dagger$ is an atomic event, whereas $t \ne t_\dagger$ is non-atomic, because it's composed of all the points in sample space that aren't the empty tree.

Let $\mathcal{T} = t_1, \dots, t_n$ denote a sequence of $n$ iid sampled trees. Let $n_\dagger = \left|\\{t \in \mathcal{T}: t=t_\dagger\\}\right|$ denote the number of sampled trees that are the empty event. Let $P_\dagger(\theta) = \mathbb{P}(t_i = t_\dagger\mid\theta)$ denote the probability of the empty event given parameters $\theta$. Now, suppose we compute the likelihood of these trees given model parameters $\theta$.

$$
\begin{eqnarray}
\mathbb{P}(\mathcal{T}\mid\theta) &=& \prod_{i=1}^n \mathbb{P}(t_i\mid\theta)\\
&=& P_\dagger(\theta)^{n_\dagger} \prod_{i: t_i\ne t_\dagger} \mathbb{P}(t_i\mid\theta)\\
&=& P_\dagger(\theta)^{n_\dagger} (1-P_\dagger(\theta))^{n-n_\dagger} \prod_{i: t_i\ne t_\dagger} p(t_i\mid\theta, t_i\ne t_\dagger),
\end{eqnarray}
$$

where $p(t_i\mid\theta, t_i\ne t_\dagger)$ is the conditional density (conditioned on non-emptiness), i.e. what we compute from the MTBD model.

The above seems to show that, if we were to include the empty events in our likelihood modeling, we would need to include some extra factors involving the empty event probability.

## Probabilistic model with stopping at $n$ successes and censored failures

Now suppose the sequence of sampled trees has failures (empty events) censored, and is stopped at $n$ successes (non-empty events). All $n$ trees are non-empty events, and an unknown number $n_\dagger$ of failures occurred at unknown locations in the sequence before it reached $n$ successes. The probability of this stopped and censored process marginalizes over the number and locations of the $n_\dagger$ failures in the previous section.

$$
\begin{eqnarray}
\mathbb{P}(\mathcal{T}\mid\theta) &=& \sum_{n_\dagger=0}^\infty {n + n_\dagger - 1 \choose n_\dagger}P_\dagger(\theta)^{n_\dagger} (1-P_\dagger(\theta))^n \prod_{i=1}^n p(t_i\mid\theta, t_i\ne t_\dagger)\\
&=& \left[\sum_{n_\dagger=0}^\infty {n + n_\dagger - 1 \choose n_\dagger}P_\dagger(\theta)^{n_\dagger} (1-P_\dagger(\theta))^n\right] \prod_{i=1}^n p(t_i\mid\theta, t_i\ne t_\dagger)\\
&=& \prod_{i=1}^n p(t_i\mid\theta, t_i\ne t_\dagger),
\end{eqnarray}
$$

where the last line follows from recognizing the factor in brackets as normalizing the negative binomial distribution (the sequence is iid, so we can model the failures as Bernoulli trials).

The above seems to show that the conditional density of nonempty trees (i.e. what we compute with an MTBD model) is all we need for modeling a sequence of iid attempts stopped at a specified number of successes, which is how we've been doing simulation.

## Emptiness with fossilization

Now we can see the subtle problem with the steps 1, 2, 3 outlined at the beginning. If $\sigma=\rho=1$, then $P_\dagger(\theta) = 0$. In other words, the empty event is a null event, because we've removed any mechanism for any lineage to be unobserved. The procedure of removing trees that don't have tips surviving to the present is invalid, because all trees are non-empty, full of detailed (albeit perhaps extinct) structure for which we were supposed to compute a conditional density with the MTBD equations. The result in the last section above would be missing factors because we shouldn't have been counting anything as empty. Even under the weaker condition of only $\sigma >0$ we will run into problems: any tree that has fossils but no survivors should not count as an empty event. The conditioning in the conditional density means non-emptiness, not necessarily sampling in the present if fossilization is allowed.

## Possible fixes

- We can prune the trees (removing extinct subtrees), and use $\sigma=0$ in our likelihoods. There is already a method in the bdms simulator that does that.
- Right now, the bdms simulator doesn't do fossilization, so one option is to add that, and then we can prune and reject samples according to _observation_, not just sampling in the present.
- With some more thought, maybe we can figure out how to compute the density conditioned on sampling in the present, even for a tree with fossils. Note that the event "sampled in the present" is a subset of the event "non-empty". Then the current sample rejection scheme could be used even with fossilization.
